src/libfetchers/tarball.cc

```diff
@@ -183,6 +184,48 @@ DownloadTarballResult downloadTarball(
     };
 }
 
+struct TarballAttrs {
+    std::string url;
+    std::optional<std::string> narHash;
+    std::optional<std::string> name;
+    std::optional<std::string> rev;
+    std::optional<int> revCount;
+    std::optional<int> lastModified;
+};
+
+static const auto tarballAttrParser = []{
+    using namespace nix::fetchers::parsers;
+    return parsers::Attrs(
+        [](
+            std::string typ,
+            std::string url,
+            std::optional<std::string> narHash,
+            // Had to make this optional to make the tests pass. fetchTree does not allow name?
+            std::optional<std::string> name,
+            std::optional<std::string> rev,
+            std::optional<int> revCount,
+            std::optional<int> lastModified
+        ) {
+            return TarballAttrs {
+                .url = url,
+                .narHash = narHash,
+                .name = name,
+                .rev = rev,
+                .revCount = revCount,
+                .lastModified = lastModified,
+            };
+        },
+        new DefaultAttr<TarballAttrs,String>("type", String {} /* TODO check either missing or "tarball" */, "type", [](auto x) { return "tarball"; }),
+        new RequiredAttr<TarballAttrs,String>("url", String {}, [](auto x) { return x.url; }),
+        new OptionalAttr<TarballAttrs,String>("narHash", String {}, [](auto x) { return x.narHash; }),
+        // new DefaultAttr<TarballAttrs,String>("name", String {}, "source", [](const TarballAttrs & x) { return x.name; }),
+        new OptionalAttr<TarballAttrs,String>("name", String {}, [](auto x) { return x.name; }),
+        new OptionalAttr<TarballAttrs,String>("rev", String {}, [](auto x) { return x.rev; }),
+        new OptionalAttr<TarballAttrs,Int>("revCount", Int {}, [](auto x) { return x.revCount; }),
+        new OptionalAttr<TarballAttrs,Int>("lastModified", Int {}, [](auto x) { return x.lastModified; })
+    );
+}();
```

Comment:
I'm debating whether this is a good solution.
Good:
- it's functional
- we can initialize a `const TarballAttrs`
- no lists of functions or such, that may slow it down (as it would be nice to have a similar solution that works for defining primops with attr based arguments etc)
- bidirectional

Bad:
- boilerplate
- too little type inference
- unparseAttr is *super* ad hoc. Doesn't have a proper interface. Just "happy" template accidents.
- making the arguments and parser codecs line up is a manual process and the error messages are useless

Perhaps a `Setting`-like alternative would be easier to use?
It would allow us to spend some template "budget" on declaring what kinds of fields they are, so that we have a variation of the `TarballAttrs` to represent a properly locked input, or a variation for the attrs that should be returned by `fetchTree`.

Reply:
Maybe a syntax like this can provide better inference, at the cost of needing another type.

```
OptionalAttr("revCount", Int {}).getter([](TarballAttrs x){ return x.revCount; })
```

Regarding the boilerplate - maybe feel quite that way if we make more use of this feature when extended with doc strings.